File: MenuBuilder/frontend/src/api/billing.ts
import client from "./client";

export interface BillingForecastMonth {
  month: string;
  amount_minor: number;
  terminal_count: number;
}

export interface BillingSummary {
  as_of: string;
  currency: string;
  monthly_base_price_minor: number;
  overdue_amount_minor: number;
  overdue_terminal_count: number;
  active_terminal_count: number;
  deactivation_scheduled_count: number;
  disabled_terminal_count: number;
  admin_disabled_terminal_count: number;
  nearest_required_payment_at: string | null;
  forecast: BillingForecastMonth[];
  billing_mode?: "standard" | "post_factum" | "cert_linked";
  min_billing_periods?: number;
  allowed_billing_periods?: string | null;
  default_selection_mode?: "only_lapsed" | "all_due" | "all";
}

export interface BillingTerminal {
  terminal_id: number;
  device_id: number;
  sn: string;
  terminal_is_active: boolean;
  license_id: number | null;
  license_expires_at: string | null;
  renewal_enabled: boolean;
  deactivation_requested_at: string | null;
  billing_status: string;
  monthly_price_minor: number;
  billing_period_months: number;
  period_price_minor: number;
  periods_due: number;
  overdue_amount_minor: number;
  next_payment_at: string | null;
  next_payment_amount_minor: number;
  projected_expires_at_after_debt_payment: string | null;
  can_deactivate: boolean;
  can_cancel_deactivation: boolean;
  can_reactivate: boolean;
  included_in_forecast: boolean;
  billing_mode?: "standard" | "post_factum" | "cert_linked";
  cert_not_valid_after: string | null;
  tenant_pin_creation_enabled: boolean;
  cert_serial: string | null;
  cert_pin_price_minor: number;
  cert_operation: string;
  cert_expiring_soon: boolean;
  cert_pin_pending: boolean;
  cert_pin_expires_at: string | null;
  address?: string | null;
  note?: string | null;
  terminal_type_id?: number;
  terminal_type_name?: string | null;
  created_at?: string | null;
}

export interface DeactivateResponse {
  terminal_id: number;
  status: string;
  works_until: string | null;
  overdue_amount_minor: number;
  included_in_forecast: boolean;
}

export interface CancelDeactivationResponse {
  terminal_id: number;
  status: string;
  renewal_enabled: boolean;
}

export interface CheckoutItemRequest {
  terminal_id: number;
  advance_periods: number;
  include_license?: boolean;
  include_cert_pin?: boolean;
}

export interface CheckoutRequest {
  items: CheckoutItemRequest[];
}

export interface CheckoutItemResponse {
  terminal_id: number;
  operation: string;
  periods_due: number;
  advance_periods: number;
  amount_minor: number;
  new_expires_at: string | null;
}

export interface CheckoutResponse {
  order_id: string;
  currency: string;
  amount_minor: number;
  payment_url: string | null;
  items: CheckoutItemResponse[];
}

export interface ReactivationCheckoutRequest {
  advance_periods: number;
}

export interface ReactivationCheckoutResponse {
  order_id: string;
  currency: string;
  amount_minor: number;
  months: number;
  new_expires_at: string;
  payment_url: string | null;
}

export async function getBillingSummary(): Promise<BillingSummary> {
  const res = await client.get("/billing/summary");
  return res.data;
}

export async function getBillingTerminals(
  status?: string,
  search?: string,
): Promise<BillingTerminal[]> {
  const params: Record<string, string> = {};
  if (status) params.status = status;
  if (search) params.search = search;
  const res = await client.get("/billing/terminals", { params });
  return res.data;
}

export async function deactivateTerminal(
  terminalId: number,
): Promise<DeactivateResponse> {
  const res = await client.post(`/billing/terminals/${terminalId}/disable`);
  return res.data;
}

export async function disableTerminal(
  terminalId: number,
): Promise<DeactivateResponse> {
  const res = await client.post(`/billing/terminals/${terminalId}/disable`);
  return res.data;
}

export async function cancelDeactivation(
  terminalId: number,
): Promise<CancelDeactivationResponse> {
  const res = await client.post(`/billing/terminals/${terminalId}/enable`);
  return res.data;
}

export async function enableTerminal(
  terminalId: number,
): Promise<CancelDeactivationResponse> {
  const res = await client.post(`/billing/terminals/${terminalId}/enable`);
  return res.data;
}

export async function createCheckout(
  data: CheckoutRequest,
): Promise<CheckoutResponse> {
  const res = await client.post("/billing/checkout", data);
  return res.data;
}

export async function createReactivationCheckout(
  terminalId: number,
  data: ReactivationCheckoutRequest,
): Promise<ReactivationCheckoutResponse> {
  const res = await client.post(
    `/billing/terminals/${terminalId}/reactivation-checkout`,
    data,
  );
  return res.data;
}

export interface ConfirmPaymentResponse {
  order_id: string;
  status: string;
  paid_at: string | null;
  items_updated: number;
}

export async function confirmPayment(
  orderId: string,
): Promise<ConfirmPaymentResponse> {
  const res = await client.post(`/billing/orders/${orderId}/confirm`);
  return res.data;
}

export interface CalculateItemResponse {
  terminal_id: number;
  operation: string;
  periods_due: number;
  advance_periods: number;
  total_periods: number;
  license_amount_minor: number;
  cert_amount_minor: number;
  total_item_amount_minor: number;
  new_expires_at: string | null;
}

export interface CalculateResponse {
  currency: string;
  total_amount_minor: number;
  license_amount_minor: number;
  cert_amount_minor: number;
  items: CalculateItemResponse[];
}

export async function calculateBilling(
  data: CheckoutRequest,
): Promise<CalculateResponse> {
  const res = await client.post("/billing/calculate", data);
  return res.data;
}

File: ProcessingBackend/backend/app/schemas/billing.py
"""Pydantic schemas for the user billing API."""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict

from app.services.billing import BillingStatus


class OrderStatus(StrEnum):
    PENDING = "pending"
    PAID = "paid"
    CANCELLED = "cancelled"
    FAILED = "failed"


class BillingForecastMonthRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    month: str
    amount_minor: int
    terminal_count: int


class BillingSummaryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    as_of: datetime
    currency: str
    monthly_base_price_minor: int
    overdue_amount_minor: int
    overdue_terminal_count: int
    active_terminal_count: int
    deactivation_scheduled_count: int
    disabled_terminal_count: int
    admin_disabled_terminal_count: int
    nearest_required_payment_at: datetime | None
    forecast: list[BillingForecastMonthRead]


class BillingTerminalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    terminal_id: int
    device_id: int
    sn: str
    terminal_is_active: bool
    license_id: int | None
    license_expires_at: datetime | None
    renewal_enabled: bool
    deactivation_requested_at: datetime | None
    billing_status: BillingStatus
    monthly_price_minor: int
    billing_period_months: int
    period_price_minor: int
    periods_due: int
    overdue_amount_minor: int
    next_payment_at: datetime | None
    next_payment_amount_minor: int
    projected_expires_at_after_debt_payment: datetime | None
    can_deactivate: bool
    can_cancel_deactivation: bool
    can_reactivate: bool
    included_in_forecast: bool
    cert_serial: str | None = None
    cert_not_valid_after: datetime | None = None
    tenant_pin_creation_enabled: bool = False
    cert_pin_price_minor: int = 0
    cert_operation: str = "primary_issue"
    cert_expiring_soon: bool = False
    cert_pin_pending: bool = False
    cert_pin_expires_at: datetime | None = None
    address: str | None = None
    note: str | None = None
    terminal_type_id: int = 0
    terminal_type_name: str | None = None
    created_at: datetime | None = None


class DeactivateTerminalResponse(BaseModel):
    terminal_id: int
    status: BillingStatus
    works_until: datetime | None
    overdue_amount_minor: int
    included_in_forecast: bool


class CancelDeactivationResponse(BaseModel):
    terminal_id: int
    status: BillingStatus
    renewal_enabled: bool


class CheckoutItemRequest(BaseModel):
    terminal_id: int
    advance_periods: int = 0  # 0, 1, or 2
    include_license: bool = True
    include_cert_pin: bool = False


class CheckoutRequest(BaseModel):
    items: list[CheckoutItemRequest]


class CheckoutItemResponse(BaseModel):
    terminal_id: int
    operation: str = "renewal"
    periods_due: int
    advance_periods: int
    amount_minor: int
    new_expires_at: datetime | None = None


class CheckoutResponse(BaseModel):
    order_id: str
    currency: str
    amount_minor: int
    payment_url: str | None
    items: list[CheckoutItemResponse]


class ReactivationCheckoutRequest(BaseModel):
    advance_periods: int = 1  # 1 or 2


class ReactivationCheckoutResponse(BaseModel):
    order_id: str
    currency: str
    amount_minor: int
    months: int
    new_expires_at: datetime
    payment_url: str | None


class BillingOrderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    org_id: int
    status: OrderStatus
    currency: str
    amount_minor: int
    provider: str | None
    provider_order_id: str | None
    payment_url: str | None
    created_at: datetime
    paid_at: datetime | None
    expires_at: datetime | None


class ConfirmPaymentResponse(BaseModel):
    order_id: str
    status: OrderStatus
    paid_at: datetime | None
    items_updated: int
